modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/client/blockstream/BlockStreamSubscription.kt

```diff
@@ -0,0 +1,123 @@
+
+    override val coroutineContext: CoroutineContext = Dispatchers.IO + SupervisorJob()
+
+    private val logger = LoggerFactory.getLogger(javaClass)
+
+    private val source: MutableMap<UUID, BlockStreamStorage> = mutableMapOf()
+    private var initialStorageId: UUID? = null
+    private var stopped: Boolean = false
+
+    fun subscribe(): Pair<UUID, BlockStreamSubscription> {
+        if (initialStorageId == null) {
+            initialStorageId = expand(context.closeIf, context.onFailure, context.onBlock)
+            launch { run() }
+        }
+        logger.info("Subscribed to block stream")
+        return initialStorageId!! to getInstance(context)
+    }
+
+    fun unsubscribe() {
+        stopped = true
+        this.cancel()
+        destroy()
+
+        logger.info("Unsubscribed from block stream")
+    }
+
+    fun expand(
+        closeIf: (suspend (block: VersionedBlockMessage) -> Boolean)? = null,
+        onFailure: (suspend (t: Throwable) -> Unit)? = null,
+        onBlock: suspend (block: VersionedBlockMessage) -> Any,
+    ): UUID {
+        val storage = BlockStreamStorage(onBlock, closeIf, onFailure ?: context.onFailure, Channel())
+        source[storage.getId()] = storage
+        logger.debug("Block stream subscription has been expanded. Number of channels is ${source.size}")
+
+        return storage.getId()
+    }
+
+    fun <T> receive(actionId: UUID, collector: FlowCollector<T>) = launch {
+        receiveBlocking<T>(actionId).collect(collector)
+    }
+
+    fun <T> receiveBlocking(actionId: UUID): Flow<T> {
+        val storage = source[actionId] ?: throw IrohaSdkException("Flow#$actionId not found")
+        return storage.channel.cast<Channel<T>>().receiveAsFlow().catch { storage.onFailure(it) }
+    }
+
+    private suspend fun run() {
+        var counter = 0
+        val request = VersionedBlockSubscriptionRequest.V1(BlockSubscriptionRequest(BigInteger.valueOf(context.from)))
+
+        context.client.webSocket(
+            host = context.apiUrl.host,
+            port = context.apiUrl.port,
+            path = Iroha2Client.WS_ENDPOINT_BLOCK_STREAM,
+        ) {
+            logger.debug("WebSocket opened")
+            send(VersionedBlockSubscriptionRequest.encode(request).toFrame())
+
+            for (frame in incoming) {
+                if (stopped) {
+                    this.close()
+                    source.closeAndClear()
+                    return@webSocket
+                }
+                logger.debug("Received frame: {}", frame)
+
+                val block = VersionedBlockMessage.decode(frame.readBytes())
+                source.forEach { (id, storage) ->
+                    val result = storage.onBlock(block)
+                    logger.debug("{} action result: {}", storage.getId(), result)
+                    storage.channel.send(result)
+
+                    if (storage.closeIf?.let { it(block) } == true) {
```

Comment:
this should be documented (`closeIf` happens after `onBlock`, not before)

Reply:
probably better to do before 🤔

Reply:
`count` param works properly if do `closeIf` after `onBlock`

Reply:
we don't need `count` at all, if you know what height you start from a `closeIf` can just contain a logic like `newBlockHeight==(x+5)`